### PR TITLE
[WIP]Refactor MulticlassMachine

### PR DESCRIPTION
**Pull request by LiuYuHui**
~~Fix by https://github.com/shogun-toolbox/shogun-libtooling/pull/2~~
- [x] fix C++ code
- [x] fix unit test 
- [x] fix meta example
- [ ] fix jupyter notebook

### CODE REVIEW

File: src/shogun/multiclass/MulticlassLibLinear.cpp [hunk 1]
```diff
@@ -55,15 +55,15 @@ MulticlassLibLinear::~MulticlassLibLinear()
 	reset_train_state();
 }
 
-SGVector<int32_t> MulticlassLibLinear::get_support_vectors() const
+SGVector<int32_t> MulticlassLibLinear::get_support_vectors() , const std::shared_ptr<Labels>& labsconst
```
Comment (gf712) [inline src/shogun/multiclass/MulticlassLibLinear.cpp]: this is really odd :D

File: src/shogun/multiclass/MulticlassLibLinear.h [hunk 2]
```diff
@@ -138,12 +138,12 @@ class MulticlassLibLinear : public RandomMixin<LinearMulticlassMachine>
 		/** get support vector indices
 		 * @return support vector indices
 		 */
-		SGVector<int32_t> get_support_vectors() const;
+		SGVector<int32_t> get_support_vectors() , const std::shared_ptr<Labels>& labsconst;
```
Comment (gf712) [inline src/shogun/multiclass/MulticlassLibLinear.h]: same here... not sure what is happening?

File: src/shogun/multiclass/ShareBoost.cpp [hunk 3]
```diff
@@ -40,15 +40,15 @@ SGVector<int32_t> ShareBoost::get_activeset()
...
+bool ShareBoost::train_machine(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& labs)
 {
 	if (data)
 		set_features(data);
 	auto fea = m_features->as<DenseFeatures<float64_t>>();
 
 	if (m_features == NULL)
 		error("No features given for training");
-	if (m_labels == NULL)
+	if (labs == NULL)
```
Comment (gf712) [inline src/shogun/multiclass/ShareBoost.cpp]:
```suggestion
	if (!labs)
```

File: src/shogun/multiclass/ShareBoost.cpp [hunk 4]
```diff
@@ -40,15 +40,15 @@ SGVector<int32_t> ShareBoost::get_activeset()
 	return m_activeset;
 }
 
-bool ShareBoost::train_machine(std::shared_ptr<Features> data)
+bool ShareBoost::train_machine(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& labs)
 {
 	if (data)
 		set_features(data);
 	auto fea = m_features->as<DenseFeatures<float64_t>>();
```
Comment (gf712) [inline src/shogun/multiclass/ShareBoost.cpp]: so `m_features` is now still part of `Machine`?

File: src/shogun/multiclass/ShareBoost.h [hunk 5]
```diff
@@ -56,13 +56,13 @@ class ShareBoost: public LinearMulticlassMachine
 protected:
 
 	/** train machine */
-	virtual bool train_machine(std::shared_ptr<Features> data = NULL);
+	virtual bool train_machine(const std::shared_ptr<Features>& data = NULL, const std::shared_ptr<Labels>& labs);
```
Comment (gf712) [inline src/shogun/multiclass/ShareBoost.h]:
```suggestion
	virtual bool train_machine(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& labs);
```

**Comment by gf712** [comment]
I guess this just has to be cleaned up a bit manually. And then can see if the compiler still throws errors, in which case there might be a class missing

File: src/shogun/classifier/svm/SVM.h [hunk 6]
```diff
@@ -64,6 +64,8 @@ class SVM : public KernelMachine
 		 */
 		SVM(float64_t C, std::shared_ptr<Kernel> k, std::shared_ptr<Labels> lab);
 
+
```
Comment (gf712) [inline src/shogun/classifier/svm/SVM.h]: extra space, I think you meant to put the space between constructor and destructor

File: src/shogun/classifier/svm/SVM.h [hunk 7]
```diff
@@ -64,6 +64,8 @@ class SVM : public KernelMachine
 		 */
 		SVM(float64_t C, std::shared_ptr<Kernel> k, std::shared_ptr<Labels> lab);
```
Comment (gf712) [inline src/shogun/classifier/svm/SVM.h:65]: do you still need this constructor?

File: src/shogun/machine/LinearMulticlassMachine.h [hunk 8]
```diff
@@ -30,20 +30,17 @@ class LinearMulticlassMachine : public MulticlassMachine
...
-		 * @param labs labels
 		 */
-		LinearMulticlassMachine(std::shared_ptr<MulticlassStrategy> strategy, std::shared_ptr<Features> features, std::shared_ptr<Machine> machine, std::shared_ptr<Labels> labs) :
-			MulticlassMachine(strategy, machine,labs)
+		LinearMulticlassMachine(std::shared_ptr<MulticlassStrategy> strategy, std::shared_ptr<Machine> machine ) :
+			MulticlassMachine(strategy, machine)
 		{
-			set_features(features->as<DotFeatures>());
-			SG_ADD(&m_features, "m_features", "Feature object.");
+
```
Comment (gf712) [inline src/shogun/machine/LinearMulticlassMachine.h]: remove space

File: src/shogun/machine/LinearMulticlassMachine.h [hunk 9]
```diff
@@ -57,38 +54,22 @@ class LinearMulticlassMachine : public MulticlassMachine
...
-
-		/** get features
-		 *
-		 * @return features
-		 */
-		std::shared_ptr<DotFeatures> get_features() const
-		{
-			return m_features;
-		}
+
```
Comment (gf712) [inline src/shogun/machine/LinearMulticlassMachine.h]: remove space

File: src/shogun/machine/LinearMulticlassMachine.h [hunk 10]
```diff
@@ -57,38 +54,22 @@ class LinearMulticlassMachine : public MulticlassMachine
...
-		std::shared_ptr<DotFeatures> get_features() const
-		{
-			return m_features;
-		}
+		
 
 	protected:
 
-		bool train_machine(std::shared_ptr<Features> data) override
+		bool train_machine(const std::shared_ptr<Features> &data, const std::shared_ptr<Labels>& labs) override
```
Comment (gf712) [inline src/shogun/machine/LinearMulticlassMachine.h]:
```suggestion
		bool train_machine(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& labs) override
```

File: src/shogun/machine/LinearMulticlassMachine.h [hunk 11]
```diff
@@ -161,23 +132,17 @@ class LinearMulticlassMachine : public MulticlassMachine
...
+			
 		}
 
 		/** deletes any subset set to the features of the machine */
 		virtual void remove_machine_subset()
 		{
-			/* changing the subset structure to use subset stacks. This might
-			 * have to be revised. Heiko Strathmann */
-			m_features->remove_subset();
+
```
Comment (gf712) [inline src/shogun/machine/LinearMulticlassMachine.h:147]: if these methods becomes noops can they just be removed? How are the subsets handled now?

File: src/shogun/machine/MulticlassMachine.cpp [hunk 12]
```diff
@@ -40,20 +38,17 @@ MulticlassMachine::~MulticlassMachine()
 {
 }
 
-void MulticlassMachine::set_labels(std::shared_ptr<Labels> lab)
-{
-    Machine::set_labels(lab);
-}
+
```
Comment (gf712) [inline src/shogun/machine/MulticlassMachine.cpp]: remove space

File: src/shogun/machine/LinearMulticlassMachine.h (hunk 11 above)
Comment (LiuYuHui) [reply src/shogun/machine/LinearMulticlassMachine.h:147]: just add/remove subset to feature that will be passed in `train_machine`.

File: src/shogun/multiclass/ShareBoost.h [hunk 13]
```diff
@@ -74,6 +71,10 @@ class ShareBoost: public LinearMulticlassMachine
 	SGMatrix<float64_t> m_rho; ///< cache_matrix for rho
 	SGVector<float64_t> m_rho_norm; ///< column sum of m_rho
 	SGMatrix<float64_t> m_pred; ///< predictions, used in training
+	std::shared_ptr<DenseFeatures<float64_t>> m_features;
+	std::shared_ptr<Labels> m_share_boost_labels;
+
+
```
Comment (gf712) [inline src/shogun/multiclass/ShareBoost.h]: remove spaces

File: src/shogun/machine/LinearMulticlassMachine.h (hunk 11 above)
Comment (gf712) [reply src/shogun/machine/LinearMulticlassMachine.h:147]: I see, makes sense to me :)

**Pull request merged**